### ISSUE TITLE
Adopt GitFlow with GitVersion

### ISSUE THREAD

**Issue by SeanFeldman** [Azure/azure-service-bus-dotnet#224]
[GitVersion](https://gitversion.readthedocs.io/en/latest/) allows easy implementation of [GitFlow](https://gitversion.readthedocs.io/en/latest/git-branching-strategies/gitflow-examples/) with SemVer in mind.

**Comment by SeanFeldman** [comment]
With the new client, will be there any support branches or will it be the "evergreen" version only?

**Comment by nemakam** [comment]
@SeanFeldman  Currently, we do not have plans to maintain support branches. If we find a bug, it will be addressed in the latest version of the code unless there are huge breaking changes. If there are, then we will look at it on a case-by-case basis.

**Comment by SeanFeldman** [comment]
While working on the repo, I've notices something something that looks off. Here's the state of `master` and `develop` branches:

![image](https://user-images.githubusercontent.com/1309622/37856881-8f08f160-2ebc-11e8-93b9-8a1aa58545a0.png)

If you notice, `master` was used to release `3.0.0-preview-02`. Meaning new development is on `master`. But `master` should represent what's currently released as a stable version, which is `2.0.0`. What's wrong with that? A few things.

1. `master` is **no longer what the latest released stable package is**. So when looking at `master` when trying to figure out a bug, there will be descrepency.
1. If there's a **need to hotfix a bug** in `2.0.0` and release a patch `2.0.1`, `master` is "contaminated" by other work in progress slotted for the next major, meaning there'd need to be yet another branch off the tag `2.0.0`. 

This is not a good path to follow. GitFlow is designed to address these issues.
`master` - always represents the latest stable released version.
`develop` - contains the new development
`release-x` - future releases that can go in parallel with hotfixes
`hotfix-y` - patches that are added to `master`, propograte into `develop` and `release-x`es.

Please review branching practice currently used.

**Comment by nemakam** [comment]
@SeanFeldman 
Having a separate branch would still be needed in GitFlow right? i.e., if we had already released 3.0 and we want to patch 2.0.1, then 2.0 would still be a new branch, right?
But that said, we should be adhering to the git standards. So we will follow git flow from next version.

**Comment by SeanFeldman** [comment]
Correct. You can brench off the `2.0.0` tag on master.
Could call it `support-2.0` and keep around. Any work would be merged into it (such `hotfix-2.0.1`).

> So we will follow git flow from next version.

![image](https://user-images.githubusercontent.com/1309622/37932519-ecd0306a-3105-11e8-8353-ce86a8262a85.png)

**Comment by SeanFeldman** [comment]
Jokes and memes aside, I trully appreciate how open-minded and listening you folks are. Kudos!

**Comment by SeanFeldman** [comment]
https://github.com/AArnott/Nerdbank.GitVersioning

**Comment by SeanFeldman** [comment]
Closing in favour of #584
